sdk/cosmos/azure_data_cosmos/src/models/patch_operations.rs

```diff
@@ -0,0 +1,323 @@
+use azure_core::Model;
+use serde::{Deserialize, Serialize};
+
+// Cosmos' patch operations are _similar_ to JSON Patch (RFC 6902) in structure, but have different operations.
+
+/// Represents a patch document for Cosmos DB.
+///
+/// A patch document is made up of a collection of patch operations.
+/// Each operation describes how to modify a property of a document.
+/// See <https://learn.microsoft.com/en-us/azure/cosmos-db/partial-document-update> for more information on patch operations.
+///
+/// # Examples
+///
+/// To build up a patch document, use [`PatchDocument::default`] to create an empty document,
+/// then use the various `with_` methods to add operations.
+///
+/// ```rust
+/// # use azure_data_cosmos::models::{PatchDocument, PatchOperation};
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let patch = PatchDocument::default()
+///     .with_add("/color".to_string(), "silver")?
+///     .with_move("/from".to_string(), "/to".to_string())?;
+/// # assert_eq!(patch, PatchDocument {
+/// #     operations: vec![
+/// #         PatchOperation::Add {
+/// #             path: "/color".to_string(),
+/// #             value: serde_json::Value::String("silver".to_string()),
+/// #        },
+/// #        PatchOperation::Move {
+/// #            from: "/from".to_string(),
+/// #            to: "/to".to_string(),
+/// #        },
+/// #    ],
+/// # });
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Model, Default, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PatchDocument {
+    pub operations: Vec<PatchOperation>,
+}
+
+impl PatchDocument {
+    /// Adds a new "add" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to add.
+    /// * `value` - The value to add.
+    pub fn with_add(
+        mut self,
+        path: String,
+        value: impl Serialize,
```

Comment:
This feels like a clear exception to our "avoid `impl` types" guidance. `Serialize` is not object safe, so we can't accept `&dyn Serialize`, and the intent here is to accept anything that can be serialized, so you can write this:

```rust
let patch = PatchDocument::default()
    .with_add("/customer", Customer::new("John Doe"));
```

Or even support any number of already-serializable types from `std`:

```rust
let patch = PatchDocument::default()
    .with_add("/customer/name", "John Doe")
    .with_add("/customer/age", 42)
    .with_add("/customer/is_enterprise", true);
```

Thoughts, @heaths ? There's also an aesthetics decision to be made about whether or not these should be `impl Trait` or an actual generic parameter. Either is OK with me, but I find `impl Trait` a little more concise while remaining clear and explicit.

Reply:
The main alternative here would be to accept `serde_json::Value` directly, which pushes that serialization to the user at the call-site.

Reply:
Again, we don't want to avoid `impl` entirely - just in places where it doesn't help or hides allocations. I think it's fine to use in many cases. Maybe we need to clarify in guidance where we don't *ideally* want it used e.g., when passed to client or client methods that are just part of URLs or a body. When used in higher-order abstractions, `impl` may be fine. It would be hard to qualify, though, so maybe we just keep it as a architectural design question per crate. @RickWinter or @JeffreyRichter, what do you think?

Reply:
I agree. The guidance is to avoid it for values related to making the HTTP request/response itself. Our guidance does not cover other values typically and I don't think we need to address it in our guidance. But perhaps make it clear that our guidelines are specifically about clients and their methods related to HTTP requests/responses.

Similarly, the .NET library guidance is about the PUBLIC face of the library and not about its internal implementation.

Reply:
Sounds good, I definitely think the `impl Trait` syntax is really helping here.

> the .NET library guidance is about the PUBLIC face of the library

I do want to make sure it's clear that this is absolutely public API, it's basically a builder API for a model type. And to that point, this is also a case where we're looking to make an exception to our "avoid builders" guideline, since I think this is exactly a situation where a builder is useful. This isn't quite a "builder" since there's no separate builder type but it has the same effect. Again, this feels like a good example of where it **does** make sense to have a builder, since the APIs for constructing a `PatchDocument` manually, while still available, are very clunky.

The above example:

```rust
let patch = PatchDocument::default()
    .with_add("/customer/name", "John Doe")
    .with_add("/customer/age", 42)
    .with_add("/customer/is_enterprise", true);
```

If written without the `with_` "consuming setters" that are effectively a built-in builder, would look like this:

```rust
let patch = PatchDocument {
    operations: vec![
        PatchOperation::Add {
            path: "/customer/name".into(),
            value: serde_json::Value::String("John Doe".into()),
        }
        // And I'm already tired of typing and don't want to finish, with two operations still to add 😛 
    ]
}
```

I think we're all on the same page there, just laying out the places where we are wiggling around the guidelines, to help demonstrate the concrete "exceptions".

Reply:
Oh, I'm afraid my analogy confused things. Forget I said anything about .NET and public apis. Our SDK Rust guidance is specifically about the client types and methods that make HTTP calls. It is NOT about other parts of the SDK that are service-specific.

Reply:
I was pretty sure you followed, I just wanted to make sure things were absolutely clear 😉